File: .release/src/discord-notify.js
import { readFileSync, existsSync } from 'fs';

function log(...args) {
  console.log(`[${new Date().toISOString()}] [Discord]`, ...args);
}

function die(...args) {
  console.error(`[${new Date().toISOString()}] FATAL:`, ...args);
  process.exit(1);
}

function env(key, fallback) {
  const val = process.env[key];
  if (val === undefined || val === '') {
    if (fallback !== undefined) return fallback;
    die(`Missing env var: ${key}`);
  }
  return val;
}

async function sendDiscordNotification() {
  log('=== Discord Notification ===');

  const webhookUrl = env('DISCORD_WEBHOOK_URL');
  const version = env('VERSION');
  const repoName = env('REPO_NAME', 'unknown');
  const configPath = env('DISCORD_CONFIG_PATH', '/tmp/discord-config.yml');

  if (!webhookUrl || webhookUrl === '') {
    log('No DISCORD_WEBHOOK_URL set, skipping');
    return;
  }

  // Build the embed payload
  const payload = {
    username: repoName,
    avatar_url: undefined,
    content: '',
    embeds: [],
  };

  let embed = {
    title: `${repoName} v${version}`,
    description: '',
    url: `https://github.com/${env('REPO_OWNER', 'castledking')}/${repoName}`,
    color: 1474606,
    timestamp: new Date().toISOString(),
    footer: { text: 'Released via GitHub Actions' },
  };

  // Read release notes
  const notesPath = `/tmp/release-notes/v${version}.md`;
  if (existsSync(notesPath)) {
    const content = readFileSync(notesPath, 'utf-8').replace(/\r\n/g, '\n');
    const lines = content.split('\n');
    if (lines[0]) embed.title = lines[0].trim();
    const body = lines.slice(1).join('\n').trim().substring(0, 4000);
    if (body) embed.description = body;
  }

  // Try reading discord-updater-webhook.yml for custom config
  if (existsSync(configPath)) {
    try {
      const yaml = readFileSync(configPath, 'utf-8');
      const config = parseSimpleYaml(yaml);

      if (config.message?.username) payload.username = config.message.username;
      if (config.message?.avatar_url) payload.avatar_url = config.message.avatar_url;
      if (config.message?.content) payload.content = config.message.content;
      if (config.embed?.color) embed.color = config.embed.color;
      if (config.embed?.footer?.text) embed.footer.text = config.embed.footer.text;
      if (config.embed?.author) {
        embed.author = {
          name: config.embed.author.name || '',
          url: config.embed.author.url || '',
          icon_url: config.embed.author.icon_url || '',
        };
      }
      if (config.embed?.thumbnail?.url) embed.thumbnail = { url: config.embed.thumbnail.url };
      if (config.embed?.image?.url) embed.image = { url: config.embed.image.url };
      if (config.embed?.timestamp === false) delete embed.timestamp;
      if (config.embed?.fields) embed.fields = config.embed.fields;
    } catch (e) {
      log('Failed to parse discord config:', e.message);
    }
  }

  payload.embeds = [embed];

  log(`Sending Discord notification for ${repoName} v${version}`);

  const response = await fetch(webhookUrl, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(payload),
  });

  if (!response.ok) {
    const text = await response.text();
    throw new Error(`Discord webhook ${response.status}: ${text}`);
  }

  log('Discord notification sent');
}

function parseSimpleYaml(text) {
  const result = {};
  let current = result;
  const path = [result];

  for (const line of text.split('\n')) {
    const trimmed = line.trim();
    if (!trimmed || trimmed.startsWith('#')) continue;

    const indent = line.search(/\S/);
    const keyMatch = trimmed.match(/^(\w[\w-]*):\s*(.*)$/);
    if (!keyMatch) continue;

    const [, key, val] = keyMatch;

    // Adjust depth based on indent
    const depth = Math.floor(indent / 2);
    while (path.length > depth + 1) path.pop();
    current = path[path.length - 1];

    if (val === '') {
      // Nested object
      const obj = {};
      current[key] = obj;
      path.push(obj);
      current = obj;
    } else if (val === '[]') {
      current[key] = [];
    } else {
      // Parse value
      let parsed = val;
      if (parsed === 'true') parsed = true;
      else if (parsed === 'false') parsed = false;
      else if (/^\d+$/.test(parsed)) parsed = parseInt(parsed, 10);
      else parsed = parsed.replace(/^"(.*)"$/, '$1');
      current[key] = parsed;
    }
  }

  return result;
}

sendDiscordNotification().catch(err => die(err.message));

File: .release/src/discord-notify.py
#!/usr/bin/env python3
"""Discord notification for plugin releases. Reads config from env vars and YAML."""
import os, sys, json
from datetime import datetime, timezone

try:
    import yaml
except ImportError:
    print("PyYAML not installed, trying pip...")
    import subprocess
    subprocess.check_call([sys.executable, "-m", "pip", "install", "pyyaml"])
    import yaml

def main():
    webhook_url = os.environ.get('DISCORD_WEBHOOK_URL', '')
    if not webhook_url:
        print("No DISCORD_WEBHOOK_URL set, skipping")
        return

    version = os.environ['VERSION']
    version_number = version[1:] if version.startswith('v') else version
    release_tag = version if version.startswith('v') else f"v{version}"
    display_version = release_tag
    repo_name = os.environ.get('REPO_NAME', 'unknown')
    repo_owner = os.environ.get('REPO_OWNER', 'castledking')

    payload = {}

    config_path = os.environ.get('DISCORD_CONFIG_PATH', '.release/discord-updater-webhook.yml')
    config = {}
    if os.path.exists(config_path):
        with open(config_path) as f:
            config = yaml.safe_load(f) or {}

    if 'message' in config:
        msg = config['message']
        if msg.get('username'): payload['username'] = msg['username']
        if msg.get('avatar_url'): payload['avatar_url'] = msg['avatar_url']
        if msg.get('content'): payload['content'] = msg['content']

    embed = {}
    use_short = config.get('use-short-md', False)
    notes_suffix = '-SHORT' if use_short else ''
    # Files live in .release/latest/ in the repo; CI copies them to /tmp/release-notes/
    # for the Modrinth/Spigot scripts. Prefer the in-repo path so Discord works without
    # the workflow having to copy/mount anything.
    notes_path = os.environ.get('RELEASE_NOTES_PATH', f".release/latest/v{version_number}{notes_suffix}.md")
    if os.path.exists(notes_path):
        with open(notes_path) as f:
            lines = f.readlines()
        embed['title'] = lines[0].strip() if lines else f"{repo_name} {display_version}"
        body = ''.join(lines[1:]).strip()
        if len(body) > 4000: body = body[:4000] + "..."
        embed['description'] = body if body else "No release notes available."
    else:
        embed['title'] = f"{repo_name} {display_version}"
        embed['description'] = "No release notes available."

    embed['url'] = f"https://github.com/{repo_owner}/{repo_name}"

    if 'embed' in config:
        ec = config['embed']
        if ec.get('color'): embed['color'] = ec['color']
        if ec.get('author'):
            embed['author'] = {
                'name': ec['author'].get('name', ''),
                'url': ec['author'].get('url', ''),
                'icon_url': ec['author'].get('icon_url', '')
            }
        if ec.get('thumbnail', {}).get('url'):
            embed['thumbnail'] = {'url': ec['thumbnail']['url']}
        if ec.get('image', {}).get('url'):
            embed['image'] = {'url': ec['image']['url']}
        if ec.get('footer'):
            embed['footer'] = {
                'text': ec['footer'].get('text', ''),
                'icon_url': ec['footer'].get('icon_url', '')
            }
        if ec.get('timestamp', True):
            embed['timestamp'] = datetime.now(timezone.utc).isoformat()
        if 'fields' in ec:
            embed['fields'] = []
            for field in ec['fields']:
                embed['fields'].append({
                    'name': field.get('name', ''),
                    'value': field.get('value', ''),
                    'inline': field.get('inline', False)
                })

    payload['embeds'] = [embed]
    payload_json = json.dumps(payload)

    import urllib.request
    # Cloudflare blocks the default Python-urllib User-Agent with rule 1010.
    # Send a bot-style UA so the webhook actually reaches Discord.
    ua = f'{repo_name}ReleaseBot/1.0 (+https://github.com/{repo_owner}/{repo_name})'

    # If edit.message_id is set, PATCH that message; otherwise POST a new one.
    # Doing both produces a duplicate embed.
    msg_id = (config.get('edit') or {}).get('message_id', '')
    if msg_id:
        url = f"{webhook_url}/messages/{msg_id}"
        method = 'PATCH'
        success_label = f"Discord message {msg_id} updated"
    else:
        url = webhook_url
        method = 'POST'
        success_label = "Discord notification sent"

    req = urllib.request.Request(
        url, data=payload_json.encode('utf-8'),
        headers={'Content-Type': 'application/json', 'User-Agent': ua},
        method=method)

    try:
        resp = urllib.request.urlopen(req)
        print(f"{success_label} (HTTP {resp.status})")
    except urllib.error.HTTPError as e:
        body = e.read().decode()
        print(f"Discord webhook error: {e.code} {body}")
        sys.exit(1)

if __name__ == '__main__':
    main()
